analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs

```diff
@@ -105,21 +121,186 @@ public void Branching_BoolOperation_LearnsBoolConstraint()
+        }
+
+        [DataTestMethod]
+        [DataRow("arg == isObject")]
+        [DataRow("isObject == arg")]
+        public void Branching_LearnsObjectConstraint_UndefinedInOtherBranch_CS(string expression)
+        {
+            var code = @$"
+object isNull = null;
+var isObject = new object();
+if ({expression})
+{{
+    Tag(""If"", arg);
+}}
+else
+{{
+    Tag(""Else"", arg);
+}}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.Should().BeNull("We can't tell if it is Null or NotNull in this branch"));
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg != isObject")]
+        [DataRow("isObject != arg")]
+        public void Branching_LearnsObjectConstraint_UndefinedInOtherBranch_Negated_CS(string expression)
+        {
+            var code = @$"
+object isNull = null;
+var isObject = new object();
+if ({expression})
+{{
+    Tag(""If"", arg);
+}}
+else
+{{
+    Tag(""Else"", arg);
+}}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateTag("If", x => x.Should().BeNull("We can't tell if it is Null or NotNull in this branch"));
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("Arg Is Nothing")]
+        [DataRow("Arg = Nothing")]
+        [DataRow("Nothing Is Arg")]
+        [DataRow("Nothing = Arg")]
+        [DataRow("Not Not Not Arg <> Nothing")]
+        [DataRow("Not Not Not Nothing <> Arg")]
+        public void Branching_LearnsObjectConstraint_VB(string expression)
+        {
+            var code = @$"
+If {expression} Then
+    Tag(""If"", Arg)
+Else
+    Tag(""Else"", Arg)
+End If
+Tag(""End"", Arg)";
+            var validator = SETestContext.CreateVB(code, ", Arg As Object").Validator;
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.Null))
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.NotNull));
```

Comment:
This is easier to read and gives better assertion messages (ordering matters):
```suggestion
            validator.TagValues("End").Should().Equal(
                new SymbolicValue().WithConstraint(BoolConstraint.True),
                new SymbolicValue().WithConstraint(BoolConstraint.False));
```
or ordering is ignored:
```suggestion
            validator.TagValues("End").Should().BeEquivalentTo(
                new SymbolicValue().WithConstraint(BoolConstraint.True),
                new SymbolicValue().WithConstraint(BoolConstraint.False));
```

Reply:
I'll keep this style. I'd like to improve the assertions framework instead to have something like
```
validator.TagValues("End").Should().ContainValuesWithExactly(BoolConstraint.True, BoolConstraint.False)
```